### ISSUE TITLE
regression test failures with --api=3.0.0

### ISSUE THREAD

**Issue by petrovr** [openssl/openssl#10224]
Build without deprecated API is fine
```
...
/usr/bin/perl ./Configure --prefix=XXXX/master_inplace shared no-ssl3 zlib-dynamic enable-unit-test --api=3.0.0 linux-x86_64
...
CFLAGS          = -Wall -O3 -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -g
...
```

```
No compiler warning with above flags.

Unfortunately now some tests start to fail:
...
Test Summary Report
-------------------
03-test_internal_bn.t            (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
03-test_internal_rsa_sp800_56b.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
10-test_bn.t                     (Wstat: 512 Tests: 7 Failed: 2)
  Failed tests:  6-7
  Non-zero exit status: 2
15-test_genrsa.t                 (Wstat: 512 Tests: 5 Failed: 2)
  Failed tests:  3, 5
  Non-zero exit status: 2
15-test_mp_rsa.t                 (Wstat: 512 Tests: 31 Failed: 2)
  Failed tests:  8, 23
  Non-zero exit status: 2
15-test_rsa.t                    (Wstat: 256 Tests: 6 Failed: 1)
  Failed test:  3
  Non-zero exit status: 1
20-test_enc_more.t               (Wstat: 256 Tests: 2 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
30-test_evp_extra.t              (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=180, Tests=1576, 200 wallclock secs (11.03 usr  0.85 sys + 158.77 cusr 23.30 csys = 193.95 CPU)
```

Dunno if above is related to failure in e_nss engine - segmentation fail when is created DSA certificate for test purpose (only OpenSSL commands).

**Comment by richsalz** [comment]
Yes, I have seen this too. :(

**Comment by levitte** [comment]
I'm seeing this too.  A quick look with VF=1 indicates that this is a BIGNUM bug, the output from `test_bn` is a loooong series of cases like this:

```
        # INFO:  @ ../master/test/bntest.c:2509
        # /home/levitte/gitwrk/openssl.net/official/master/test/recipes/10-test_bn_data/bngcd.txt:2936: Failed GCD test
        # ERROR:  @ ../master/test/bntest.c:136
        # unexpected gcd(A,B) value
        # 
        # ERROR: (BIGNUM) 'expected == actual' failed @ ../master/test/bntest.c:137
        # --- expected
        # +++ actual
        #                                                               bit position
        # -                                      100000000000 0000000000000000:  256
        # -0000000000000000 0000000000000000 0000000000000000 0000000000000001:    0
        # +                                                                  0:    0
        #                                                                    ^
        # 
        # INFO:  @ ../master/test/bntest.c:2509
        # /home/levitte/gitwrk/openssl.net/official/master/test/recipes/10-test_bn_data/bngcd.txt:2940: Failed GCD test
        # ERROR:  @ ../master/test/bntest.c:136
        # unexpected gcd(A,B) value
        # 
        # ERROR: (BIGNUM) 'expected == actual' failed @ ../master/test/bntest.c:137
        # --- expected
        # +++ actual
        #                                                               bit position
        # -                                      1fffffffffff ffffffffffffffff:  256
        # -ffffffffffffffff ffffffffffffffff ffffffffffffffff ffffffffffffffff:    0
        # +                                                                  0:    0
        #                                                                    ^
        # 
```

**Comment by levitte** [comment]
I suspect the #10122 changes need deeper scrutiny

**Comment by levitte** [comment]
A quick bisect shows that the test_bn failures came with b75d6310857bc44ef2851bde68a1979c18bb4807

**Comment by levitte** [comment]
The fault is introduced in f3c4adfc7eb13e9eff514039b4c60b457bdba433, which is part of #10122.

**Comment by romen** [comment]
This issue is partially fixed on `master` with:
- 8aca4bfe82 Update control logic for BN_gcd (from #10232) 

The fix is only partial because `no-deprecated` builds still fail elsewhere due to some commit merged after #10122 was merged.

```
Test Summary Report
-------------------
20-test_enc_more.t               (Wstat: 256 Tests: 2 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=180, Tests=1574, 91 wallclock secs ( 2.98 usr  0.40 sys + 91.84 cusr  8.52 csys = 103.74 CPU)
Result: FAIL
```

**Comment by t8m** [comment]
That's from #9912 - the test_enc_more needs to be fixed to use -list instead of -ciphers.

**Comment by petrovr** [comment]
Unfortunately my expectation that correction in BN will resolve bus error was not confirmed :( .
Regression test pass without error but bus error remain:
```
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [XX]:State or Province Name (full name) [World]:Locality Name (eg, city) [Somewhere cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω]:Organization Name (eg, company) [E_NSS Test Team cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω]:Organizational Unit1 Name (eg, section1 - optional) [E_NSS Testers cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω-1]:Organizational Unit2 Name (eg, section2 - optional) [E_NSS Testers cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω-2]:Organizational Unit3 Name (eg, section3 - optional) [E_NSS Testers cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω-3]:Common Name (eg, YOUR name) []:Email Address (optional) [email@not.set]:
Please enter the following 'extra' attributes
to be sent with your certificate request
A challenge password []:<<<builddir>>>/tests/ca/ca-cre-certs.sh: line 169: 12393 Done                    echo_req_dn "E_NSS Certificate ($N)" "$GET_PASS"
     12394 Segmentation fault      | $OPENSSL req -new -utf8 -config "$CA_CONFIG" -key ../$N-priv.pem -passin $keypass -out $TMPDIR/$N.csr
```

Out of e_nss engine it could be reproduced with:
```bash
./util/opensslwrap.sh req -new -utf8 -config test/CAssdsa.cnf -key test/testdsa.pem
```
result is:
```
Can't load ./.rnd into RNG
00:F7:2D:A6:C5:7F:00:00:error:random number generator:RAND_load_file:Cannot open file:crypto/rand/randfile.c:98:Filename=./.rnd
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [ES]:ES
Organization Name (eg, company) []:Hermanos Locos
Common Name (eg, YOUR name) []:Hermanos Locos CA
Нарушение на разделянето(segfault)
```

I'm afraid that OpenSSL regression test does not catch anything used in real world:
PKIS-SSH: there is no way to pass cipher tests (issue #9524 from just one now is for other s well)
E_NSS engine: it is clear that currently externally managed DSA does not work (issue #9915) but segmentation fault is new one.

I saw some activities to improve documentation(manual pages). This is fine but in situation where basic functionality of cryptografic library is totally changed (switch to provider) is seems to me improved regression tests suite is more important.

**Comment by romen** [comment]
```
Starting program: /tmpram/failures/apps/openssl req -new -utf8 -config test/CAssdsa.cnf -key test/testdsa.pem
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Can't load ./.rnd into RNG
C0:E0:FC:F7:FF:7F:00:00:error:random number generator:RAND_load_file:Cannot open file:../../home/tuveri/repos/ossl_committer/workbench/crypto/rand/randfile.c:98:Filename=./.rnd                                                                                                                                     
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [ES]:ES
Organization Name (eg, company) []:Hermanos Locos
Common Name (eg, YOUR name) []:Hermanos Locos CA

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff76eb590 in EVP_MD_name (md=0x0) at ../../home/tuveri/repos/ossl_committer/workbench/crypto/evp/evp_lib.c:601
601         if (md->prov != NULL)
```

The latter segfault seems to be coming from https://github.com/openssl/openssl/blob/f702256d5514e54138279fd4637353e328262817/crypto/evp/evp_lib.c#L599-L608

which is called from https://github.com/openssl/openssl/blob/f702256d5514e54138279fd4637353e328262817/crypto/evp/m_sigver.c#L104-L107
which was modified as part of d8c98d79d1bb1dc3261d9b4fc2bc36074aec88f6 (from #10013)

**Comment by romen** [comment]
@mattcaswell I was looking at this and it seems like the _non-legacy_ behaviour introduced by d8c98d7 is missing the logic that would allow the input parameter `type` to be `NULL` and be correctly handled.

The new logic eventually crashes with a segfault if `type` is `NULL`, while the _legacy_ code queries the `EVP_PKEY` for the default digest

https://github.com/openssl/openssl/blob/f702256d5514e54138279fd4637353e328262817/crypto/evp/m_sigver.c#L143-L156

**Comment by romen** [comment]
On the other hand, the segfault does not depend anymore on `--api=3.0.0` or not, so we should probably tackle this as a separate issue and consider the issue originally reported fixed by #10232 and #10242 .

@petrovr would you agree with this assesment?
Would you mind closing this issue and opening a new one  based on https://github.com/openssl/openssl/issues/10224#issuecomment-546593113 ?

**Comment by romen** [comment]
See #10312 that should trigger the failure reported by @petrovr in the CI

**Comment by petrovr** [comment]
Closing as regression test are fixed and this match subject of issue.
The real reason to open this was failure in "req" command but this now is in #10312.
Thank for you work.
